src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Services/PackageSearchMetadataCacheObject.cs

```diff
@@ -0,0 +1,63 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+#nullable enable
+
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using NuGet.Protocol;
+using NuGet.Protocol.Core.Types;
+using NuGet.VisualStudio.Internal.Contracts;
+
+namespace NuGet.PackageManagement.VisualStudio
+{
+    internal sealed class PackageSearchMetadataCacheObject
+    {
+        private readonly IPackageSearchMetadata _packageSearchMetadata;
+        private readonly IPackageMetadataProvider _packageMetadataProvider;
+
+        public PackageSearchMetadataCacheObject(IPackageSearchMetadata packageSearchMetadata, IPackageMetadataProvider packageMetadataProvider)
+        {
+            _packageSearchMetadata = packageSearchMetadata;
+            _packageMetadataProvider = packageMetadataProvider;
+            AllVersionsContextInfo = GetVersionInfoContextInfoAsync();
+            PackageDeprecrationMetadataContextInfo = GetPackageDeprecationMetadataContextInfoAsync();
+            DetailedPackageSearchMetadataContextInfo = GetDetailedPackageSearchMetadataContextInfoAsync();
+        }
+
+        public Task<IReadOnlyCollection<VersionInfoContextInfo>> AllVersionsContextInfo { get; }
+        public Task<PackageDeprecationMetadataContextInfo?> PackageDeprecrationMetadataContextInfo { get; }
+        public Task<PackageSearchMetadataContextInfo> DetailedPackageSearchMetadataContextInfo { get; }
+
+        public static string GetCacheId(string packageId, bool includePrerelease, IReadOnlyCollection<PackageSourceContextInfo> packageSources)
+        {
+            string packageSourcesString = string.Join(" ", packageSources.Select(ps => ps.Name));
+            return string.Concat(packageId, ":", includePrerelease, " - ", packageSourcesString);
+        }
```

Comment:
I'm not crazy about routinely making bigger strings out of littler strings as an implementation detail.  We already UI delay / GC pause bugs from doing it elsewhere.

Is it reasonable to use hash codes instead? `HashCodeCombiner` might work.

Reply:
I was also not a huge fan of this, I will look into HashCodeCombiner.